src/Arduino_NotecardConnectionHandler.cpp

```diff
@@ -0,0 +1,979 @@
+}
+
+/******************************************************************************
+   PROTECTED STATE MACHINE FUNCTIONS
+ ******************************************************************************/
+
+NetworkConnectionState NotecardConnectionHandler::update_handleInit()
+{
+  NetworkConnectionState result = NetworkConnectionState::INIT;
+#if defined(LOG_MEMORY_USAGE)
+  logMemoryUsage(__FUNCTION__, true);
+#endif
+#if defined(STLINK_DEBUG)
+  // Output Notecard logs to the STLINK serial port
+  stlinkSerial.end();  // necessary to handle multiple initializations (e.g. reconnections)
+  stlinkSerial.begin(115200);
+  const size_t usb_timeout_ms = 3000;
+  for (const size_t start_ms = millis(); !stlinkSerial && (millis() - start_ms) < usb_timeout_ms;);
+  _notecard.setDebugOutputStream(stlinkSerial);
+#endif
+
+  // Initialize the Notecard based on the configuration
+  if (_serial) {
+    _notecard.begin(*_serial, _uart_speed);
+  } else {
+    _notecard.begin(_i2c_address, _i2c_max, *_wire);
+  }
+
+  // Configure the ATTN pin to be used as an interrupt to indicate when a Note
+  // is available to read. `getNote()` will only arm the interrupt if no old
+  // Notes are available. If `ATTN` remains unarmed, it signals the user
+  // application that outstanding Notes are queued and need to be processed.
+  if (J *note = getNote(false)) {
+    JDelete(note);
+  }
+
+  // Set the project UID
+  if (NetworkConnectionState::INIT == result) {
+    if (configureConnection(true)) {
+      result = NetworkConnectionState::INIT;
+    } else {
+      result = NetworkConnectionState::ERROR;
+    }
+  }
+
+#if defined(ARDUINO_OPTA)
+  // The Opta Extension has an onboard Li-Ion capacitor, that can be utilized
+  // to monitor the power state of the device and automatically report loss of
+  // power to Notehub. The following command enables that detection by default
+  // for the Opta Wirelss Extension.
+  if (NetworkConnectionState::INIT == result) {
+    if (J *req = _notecard.newRequest("card.voltage")) {
+      JAddStringToObject(req, "mode", "lipo");
+      JAddBoolToObject(req, "alert", true);
+      JAddBoolToObject(req, "sync", true);
+      JAddBoolToObject(req, "usb", true);
+      if (J *rsp = _notecard.requestAndResponse(req)) {
+        // Check the response for errors
+        if (NoteResponseError(rsp)) {
+          const char *err = JGetString(rsp, "err");
+          Debug.print(DBG_ERROR, F("%s"), err);
+          result = NetworkConnectionState::ERROR;
+        } else {
+          result = NetworkConnectionState::INIT;
+        }
+        JDelete(rsp);
+      } else {
+        Debug.print(DBG_ERROR, F("Failed to receive response from Notecard."));
+        result = NetworkConnectionState::ERROR; // Assume the worst
+      }
+    } else {
+      Debug.print(DBG_ERROR, "Failed to allocate request: card.voltage");
+      result = NetworkConnectionState::ERROR; // Assume the worst
+    }
+  }
+#endif
+
+#if defined(BOARD_HAS_SECRET_KEY)
```

Comment:
`BOARD_HAS_SECRET_KEY` define comes from ArduinoIoTCloud library, so I think also this piece of code should be moved there.

Reply:
The use of `BOARD_HAS_SECRET_KEY` was an optimization, it is not strictly necessary for this operation, so I have removed it.